File: LeetCode/Python3/Problem 146.py
# LeetCode: Problem 146
# Jonathan Kosasih

#Design and implement a data structure for Least Recently Used (LRU) cache.
#It should support the following operations: get and put.
#
#get(key) - Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.
#put(key, value) - Set or insert the value if the key is not already present.
#When the cache reached its capacity, it should invalidate the least recently used item before inserting a new item.
#
#The cache is initialized with a positive capacity.
#------------------------------------------------------------------
from collections import deque
class LRUCache(object):
    
    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.cache = dict()
        self.order = deque()
        

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if(key in self.cache):
            value = self.cache[key]
            self.order.remove(key)
            self.order.append(key)
            return value
        else:
            return -1

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if(key in self.cache):
            self.order.remove(key)
        self.cache[key] = value
        self.order.append(key)
        if(len(self.cache) > self.capacity):
            oldest = self.order.popleft()
            del self.cache[oldest]
#-------------------------------------------------
# using an OrderedDict, faster
from collections import OrderedDict
    def __init__(self, capacity):
        self.capacity = capacity
        self.cache = OrderedDict()

    def get(self, key):
        if(key in self.cache):
            value = self.cache[key]
            del self.cache[key]
            self.cache[key] = value
            return value
        else:
            return -1

    def put(self, key, value):
        if(key in self.cache):
            del self.cache[key]
        self.cache[key] = value
        if(len(self.cache) > self.capacity):
            self.cache.popitem(False)

File: LeetCode/C++/Problem 146.cpp
// LeetCode: Problem 146
// Jonathan Kosasih
/*
Design and implement a data structure for Least Recently Used (LRU) cache. It should support the following operations: get and put.

get(key) - Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.
put(key, value) - Set or insert the value if the key is not already present. When the cache reached its capacity, it should invalidate the least recently used item before inserting a new item.

The cache is initialized with a positive capacity.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
#include <numeric>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;
using std::accumulate;

class LRUCache {
public:
	LRUCache(int capacity) : myCap(capacity) {

	}

	int get(int key) {
		if (myMap.find(key) == myMap.end())
			return -1;
		auto it = myMap[key];
		const int val = it->second;
		myList.erase(it);
		myMap[key] = myList.insert(myList.begin(), make_pair(key, val));
		return val;
	}

	void put(int key, int value) {
		if (myMap.find(key) != myMap.end()) {
			myList.erase(myMap[key]);
			myMap.erase(key);
		}
		if (myMap.size() >= myCap) {
			int keyToBeRemoved = myList.rbegin()->first;
			myList.pop_back();
			myMap.erase(keyToBeRemoved);
		}
		myMap[key] = myList.insert(myList.begin(), make_pair(key, value));
	}
private:
	list<pair<int, int> > myList;
	unordered_map<int, list<pair<int, int> >::iterator > myMap;
	const int myCap;
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
int main()
{
    //std::cout << "Hello World!\n";
}